toolchain/check/import.cpp

```diff
@@ -87,10 +72,56 @@ static auto CacheCopiedNamespace(
       << to_scope_id;
 }
 
+// Copies a namespace from the import IR, returning its ID. This may diagnose
+// name conflicts, but that won't change the result because namespaces supersede
+// other names in conflicts.
+static auto CopySingleNameScopeFromImportIR(
+    Context& context,
+    llvm::DenseMap<SemIR::NameScopeId, SemIR::NameScopeId>& copied_namespaces,
+    SemIR::NameScopeId import_scope_id, SemIR::NameScopeId enclosing_scope_id,
+    SemIR::NameId name_id, SemIR::TypeId namespace_type_id)
+    -> SemIR::NameScopeId {
+  auto& scope = context.name_scopes().Get(enclosing_scope_id);
+  auto [it, success] = scope.names.insert({name_id, SemIR::InstId::Invalid});
+  if (!success) {
+    if (auto namespace_inst =
+            context.insts().TryGetAs<SemIR::Namespace>(it->second)) {
+      // Namespaces are open, so we can append to the existing one even if it
+      // comes from a different file.
+      CacheCopiedNamespace(copied_namespaces, import_scope_id,
+                           namespace_inst->name_scope_id);
+      return namespace_inst->name_scope_id;
+    }
+  }
+
+  // Produce the namespace for the entry. Use the invalid node because there's
+  // no node to associate with.
```

Comment:
Fine to use the invalid node in this iteration, but you could also use the `import` parse node maybe?

Reply:
I think the import would be something belonging in a note; it's not the name conflict and would be misleading.

Reply:
And to be clear, adding it I think would take us further from a correct implementation.